cli/test/smokehouse/test-definitions/a11y.js

```diff
@@ -503,6 +503,11 @@ const expectations = {
           ],
         },
       },
+      // TODO(jasmineyan): why is this not applicable?
```

Comment:
I added a test case to `a11y_tester.html` for `frame-focusable-content` that should be a failing case based on the axe docs (https://dequeuniversity.com/rules/axe/4.7/frame-focusable-content#:~:text=Incorrect%20markup%20solutions). 

But strangely it's registering as not applicable... I tried using `src` instead of `srcdoc`, and checked if removing `<section>` did anything, but no success there. I also verified that the button in the iframe is indeed not focusable when I brought up http://localhost:10503/a11y/a11y_tester.html, and is focusable by default without tabindex=-1 🤔. Any ideas appreciated here, maybe I'm missing something totally obvious.

Reply:
I hit this too: https://github.com/GoogleChrome/lighthouse/pull/12536#discussion_r637452264 `src=path/to/index.html` should work, `srcdoc=<inline html>` didn't (perhaps I was mistaken and neither worked)

Later in the thread brendan mentions that he couldn't get anything to fail this audit. I'm pasting the axe snippet we use and am seeing the same. First glance, it looks like axe is trying to send a "command" to an instance of axe it expects to be inside an iframe... but we only inject axe into the main document! https://github.com/dequelabs/axe-core/blob/1d1af0936144e342d231ff9fba348066632cf9e1/lib/core/utils/collect-results-from-frames.js#L26

Reply:
I just tested with the axe extension and it called this out as expected 😕 

So somehow we're not surfacing it.

Reply:
I'll work on adding support for this 👍

Reply:
Support for what?

Reply:
Adding axe to iframes in the main document, so axe has a way to gather the info it needs for frame-focusable-content.

Reply:
So it's definitely possible to change our gatherer to allow `frame-focusable-content` to work. It's pretty hairy, though.

branch: https://github.com/GoogleChrome/lighthouse/tree/hacky-axe-frames

- gotta inject the axe source into every iframe target. For this, I just iterate over all the execution contexts in target manager and do runtime.evaluate. We need to introduce a new api to target manager to "Give me the iframe execution contexts" - right now I just do _all_ of them. Also should probably limit to just iframes from the main document (no nested iframes)
- axe configuration `application` name _must_ be the default to work, otherwise postMessages are just ignored
- since results from an iframe go through postMessage, we lost info on node details. To work around that I added a hack to call `getNodeDetails` within the axe code, and grab the resulting object when we construction the artifact
- we don't want to surface every a11y issue in iframes (very likely to be unfixable by user...), so I hacked the message that axe sends to iframes to run itself - usually it sends the whole options object. Instead the hacked version sends a config for _just_ `frame-focusable-content`.
- can confirm things work in the branch by running on `a11y_tester_framed.html` and looking at the Accessibility artifact. You'll see:

<img width="873" alt="image" src="https://github.com/GoogleChrome/lighthouse/assets/4071474/2f380949-c76e-4f34-84b4-1597baf87f57">

any solution here is going to require rather hacky modifications to the axe source code, so if we _really_ want this audit we'll have to maintain a minor soft-fork. See the `axe-tmp.js` in my branch.

Reply:
Yeah, I don't know if the ROI is there. You'd need not just iframe execution contexts, but a canonical context so you aren't accidentally running multiple times per frame, iframes bring new CSP and same-origin script execution issues, `devtoolsNodePath`s [don't currently work across iframe boundaries](https://github.com/GoogleChrome/lighthouse/blob/87f814e8b62c621afdbd4e348543f84ec8a2e5a4/core/lib/page-functions.js#L249) (though if the node was just the iframe that could work), even limiting to the single audit we'd have at least the perf overhead of building the axe virtual DOM per frame, etc.

Seems like there are three options:
- continue to not include the audit
- hack time: to add another level to the hackiness, we could even Runtime.evaluate _just_ a [focusable content check](https://github.com/dequelabs/axe-core/blob/c806e577a6b5d1a8ceebee8c5edef38ad49fceef/lib/checks/keyboard/frame-focusable-content-evaluate.js) per iframe and then dress it up to look like a postMessaged result that the main frame's axe-core could merge together
- heavier weight: the axe team maintains libraries for major testing frameworks themselves: https://github.com/dequelabs/axe-core-npm/. We could just use `@axe-core/puppeteer`, which already handles iframes. Still have to deal with all the non-execution-context issues.

Reply:
Thanks all for looking into this! Decided to not include `frame-focusable-content` given how involved it would be.